### ISSUE TITLE
Style/PercentQLiterals: Enforce upper_case_q

### ISSUE THREAD

**Issue by jpickwell** [testdouble/standard#157]
For the `Style/PercentQLiterals` cop, the enforced style should be `upper_case_q` instead of `lower_case_q`. This is for the same reason as using `double_quotes` for `Style/StringLiterals`.

From the [Ruby documentation](https://ruby-doc.org/core-2.7.0/doc/syntax/literals_rdoc.html#label-Strings):
> There are two different types of `%` strings `%q(...)` behaves like a single-quote string (no interpolation or character escaping), while `%Q` behaves as a double-quote string.

**Comment by twe4ked** [comment]
What's the difference between `%{…}`  and `%Q{…}`? I don't see any docs for it.

```
irb(main):001:0> %{x#{1}}
=> "x1"
irb(main):002:0> %q{x#{1}}
=> "x\#{1}"
irb(main):003:0> %Q{x#{1}}
=> "x1"
```

👍 to enforcing something for consistency though.

**Comment by jpickwell** [comment]
They're the same.

I just noticed that `Style/BarePercentLiterals` is set to `bare_percent`, so `Style/PercentQLiterals` is essentially pointless.

**Comment by searls** [comment]
Thanks for the conversation and research here. I just decided to toodle around with this a bit and (as someone who in 15 years has never used _literally any_ of this syntax in practice), I was unpleasantly surprised by standard's behavior.

Examples:

```ruby
puts %w{a b} # Error: Style/PercentLiteralDelimiters: `%w`-literals should be delimited by `[` and `]`.

puts %W[a b] # Error: Style/RedundantCapitalW: Do not use `%W` unless interpolation is needed. If not, use `%w`.

puts %W{a b} # Error: Style/RedundantCapitalW: Do not use `%W` unless interpolation is needed. If not, use `%w`.

puts %Q{a b} # 4 Errors lol:
  # Style/BarePercentLiterals: Use `%` instead of `%Q`.
  # Style/PercentLiteralDelimiters: `%Q`-literals should be delimited by `(` and `)`.
  # Style/PercentQLiterals: Do not use `%Q` unless interpolation is needed. Use `%q`.
  # Style/RedundantPercentQ: Use `%Q` only for strings that contain both single quotes and double quotes, or for dynamic strings that contain double quotes.
```

With `--fix`, the expression `%Q{a b}` will first be converted to `%q{a b}` (thanks to `Style/PercentQLiterals`) and then `%{a b}` (thanks to `Style/BarePercentLiterals`) and then `%(a b)` (thanks to `Style/PercentLiteralDelimiters`). 

This all seems ridiculous.

The configuration we have for preferred delimiters is literally just rubocop's default because I'd never thought about it before and nobody else had expressed an opinion, but at a glance it looks silly that they're all different for (probably) arbitrary reasons:

```yaml
Style/PercentLiteralDelimiters:
  Enabled: true
  PreferredDelimiters:
    default: ()
    '%i': '[]'
    '%I': '[]'
    '%r': '{}'
    '%w': '[]'
    '%W': '[]'
```

I don't know what to do about any of this, but this patch seems like a minimally invasive first step to clean things up a bit:

```patch
diff --git a/config/base.yml b/config/base.yml
index bf85041..3d8ba15 100644
--- a/config/base.yml
+++ b/config/base.yml
@@ -903,10 +903,6 @@ Style/PercentLiteralDelimiters:
     '%w': '[]'
     '%W': '[]'
 
-Style/PercentQLiterals:
-  Enabled: true
-  EnforcedStyle: lower_case_q
-
 Style/PreferredHashMethods:
   Enabled: true
   EnforcedStyle: short
@@ -1048,9 +1044,6 @@ Style/TrivialAccessors:
 Style/UnlessElse:
   Enabled: true
 
-Style/RedundantCapitalW:
-  Enabled: true
-
 Style/RedundantCondition:
   Enabled: true
```

**Comment by searls** [comment]
Released the above in a quick patch release (0.1.8)

**Comment by jpickwell** [comment]
@searls, for `%i`/`%I` (symbol array) and `%w`/`%W` (string array), `[]` are used as the delimiters because they reflect that these are array literals, so that makes a lot of sense.

**Comment by drbrain** [comment]
TL;DR: I'm not sure why `%r` would use the paired delimiters `<>`, `()`, `{}`, or `[]` as you trade not needing to escape `/` for potential confusion when you don't count your open/closes correctly when matching any of the paired delimiters. This is worse if you have spaces in your Regexp. I think the non-special charter `%` is the best choice.

Let's say we want to determine the username of some path that references a home directory. We have a few options:

* `//`: `/^\/(Users|home)\/(.*?)/`
* `%r()`: `%r(/(Users|home)/(.*))`
* `%r{}`: `%r{/(Users|home)/(.*)}`
* `%r[]`: `%r[/(Users|home)/(.*)]`
* `%r%%`: `%r%/(Users|home)/(.*)%`
* `%r  `: `%r /(Users|home)/(.*) ` (\u0020, space, no I'm not serious)

The `//` Regexp forces you to escape `/` when used in the pattern which forces you to apply unescaping rules in order to read the regular expression.  This is probably the number one reason people reach for `%r` literal.

The `%r()` is better in that you don't have to escape `()`, but could cause confusion if you typo because you didn't count parentheses correctly.  The same applies to the `[]`, `{}`, and `<>` variants if you use a Regexp character class (`/[…]/`) or repetition (`/(pattern){1,3}/`), or named capture (`/(?<name>pattern)/` respectively

Using a non special character like `%` is better as there's no possibility for confusion with a special character. It also is a variant of `/` (perhaps even a Zalgo'd `/`). Another option is `~` but it looks a little weird to me, even though it makes you think of matching: `%r~/(Users|home)/([e].*){1}~ =~ Dir.pwd`

Other options include:

* `~` could work
* `!$&&+,-.:?^|` are all used in patterns
* `"'_` are too common in patterns
* `#` is used in `//x` mode for adding comments
* ``;=` `` are too weird (and space is used with `//x` to make your Regexp more readable)
* `\` is a syntax error

**Comment by searls** [comment]
@drbrain - I think you've convinced me that `%r%%` is the best delimiter for regex. How common a meme is this among people who use `%r`?

**Comment by bbatsov** [comment]
While I agree with the sentiment, I feel this can produce some pretty confusing code (especially for newcomers). We'll have to dig through GitHub, but I definitely haven't seen much usage of `%r%%` (and non-paired delimiters in general).

**Comment by pointlessone** [comment]
For `%r`, I believe, `{}` is used as it's the least used brace in regexps but it still signifies embraced expression. I've ever seen`%` used as a delimiter in sed and even then not very often.

**Comment by twe4ked** [comment]
I usually switch to `'` or `"` if `{` is being used already. I'm not usually matching both types of quotes.
